bundles/automation/org.eclipse.smarthome.automation.providers/src/main/java/org/eclipse/smarthome/automation/core/provider/ModuleTypeResourceBundleProvider.java

```diff
@@ -0,0 +1,205 @@
+            moduleTypesTracker = null;
+            moduleTypeRegistry = null;
+        }
+        super.close();
+    }
+
+    /**
+     * @see ModuleTypeProvider#getModuleType(java.lang.String, java.util.Locale)
+     */
+    public ModuleType getModuleType(String UID, Locale locale) {
+        Localizer l = null;
+        synchronized (providedObjectsHolder) {
+            l = providedObjectsHolder.get(UID);
+        }
+        if (l != null) {
+            ModuleType mt = (ModuleType) l.getPerLocale(locale);
+            return mt;
+        }
+        return null;
+    }
+
+    /**
+     * @see ModuleTypeProvider#getModuleTypes(java.util.Locale)
+     */
+    public Collection<ModuleType> getModuleTypes(Locale locale) {
+        ArrayList moduleTypesList = new ArrayList();
+        synchronized (providedObjectsHolder) {
+            Iterator i = providedObjectsHolder.values().iterator();
+            while (i.hasNext()) {
+                Localizer l = (Localizer) i.next();
+                if (l != null) {
+                    ModuleType mt = (ModuleType) l.getPerLocale(locale);
+                    if (mt != null)
+                        moduleTypesList.add(mt);
+                }
+            }
+        }
+        return moduleTypesList;
+    }
+
+    /**
+     * @see AbstractResourceBundleProvider#addingService(ServiceReference)
+     */
+    @Override
+    public Object addingService(ServiceReference reference) {
+        if (reference.getProperty(Parser.PARSER_TYPE).equals(Parser.PARSER_MODULE_TYPE)) {
+            return super.addingService(reference);
+        }
+        return null;
+    }
+
+    /**
+     * @see AbstractResourceBundleProvider#importData(Vendor, Parser, java.io.InputStreamReader, java.util.ArrayList)
+     */
+    @Override
+    protected Set<Status> importData(Vendor vendor, Parser parser, InputStreamReader inputStreamReader,
+            ArrayList<String> portfolio) {
+        if (vendor != null) {
+            synchronized (providerPortfolio) {
+                providerPortfolio.put(vendor, portfolio);
+            }
+        }
+        Set<Status> providedObjects = parser.importData(inputStreamReader);
+        if (providedObjects != null && !providedObjects.isEmpty()) {
+            Iterator i = providedObjects.iterator();
+            while (i.hasNext()) {
+                Status status = (Status) i.next();
+                ModuleType providedObject = (ModuleType) status.getResult();
+                if (providedObject != null) {
+                    String uid = providedObject.getUID();
+                    if (checkExistence(uid, status))
+                        continue;
+                    portfolio.add(uid);
+                    Localizer lProvidedObject = new Localizer(providedObject);
+                    synchronized (providedObjectsHolder) {
+                        providedObjectsHolder.put(uid, lProvidedObject);
+                    }
+                    add(providedObject);
```

Comment:
Once this NPE fixed, the next problem shows up: As the provider persists its data, on a second launch, the add() method will fail with an IllegalArgumentException() since an entry with the same key already exists in the database. This exception needs to be caught and handled.